### PR TITLE
Async n-step q-learning and one step sarsa

### PR DESCRIPTION
**Pull request by ShangtongZhang**

### CODE REVIEW

File: src/mlpack/methods/reinforcement_learning/async_learning.hpp [hunk 1]
```diff
@@ -137,6 +139,22 @@ template <
 >
 class OneStepQLearningWorker;
 
+template <
+  typename EnvironmentType,
```
Comment (zoq) [inline src/mlpack/methods/reinforcement_learning/async_learning.hpp]: Can you comment on the template parameter?

**Comment by ShangtongZhang** [comment]
Would there be more comments on this PR?

File: src/mlpack/methods/reinforcement_learning/worker/n_step_q_learning_worker.hpp [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+namespace mlpack {
+namespace rl {
+
+/**
+ * N step Q-Learning worker.
+ *
+ * @tparam EnvironmentType The type of the reinforcement learning task.
+ * @tparam NetworkType The type of the network model.
+ * @tparam UpdaterType The type of the optimizer.
+ * @tparam PolicyType The type of the behavior policy. *
```
Comment (zoq) [inline src/mlpack/methods/reinforcement_learning/worker/n_step_q_learning_worker.hpp]: Can you remove the ``*`` at the end?

File: src/mlpack/methods/reinforcement_learning/worker/n_step_q_learning_worker.hpp [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+  typename PolicyType
+>
+class NStepQLearningWorker
+{
+ public:
+  using StateType = typename EnvironmentType::State;
+  using ActionType = typename EnvironmentType::Action;
+  using TransitionType = std::tuple<StateType, ActionType, double, StateType>;
+
+  /**
```
Comment (zoq) [inline src/mlpack/methods/reinforcement_learning/worker/n_step_q_learning_worker.hpp]: Do you mind to add a method description here; something like should do:

```
Construct N step Q-Learning worker with the given parameters and environment.
```

File: src/mlpack/methods/reinforcement_learning/worker/n_step_q_learning_worker.hpp [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+        // Compute the training target for current state.
+        network.Forward(std::get<0>(transition).Encode(), actionValue);
+        actionValue[std::get<1>(transition)] = target;
+
+        // Compute gradient.
+        arma::mat gradients;
+        network.Backward(actionValue, gradients);
+
+        // Accumulate gradients.
+        totalGradients += gradients;
```
Comment (zoq) [inline src/mlpack/methods/reinforcement_learning/worker/n_step_q_learning_worker.hpp]: We should initialize  ``totalGradients`` with zero.

File: src/mlpack/methods/reinforcement_learning/worker/one_step_sarsa_worker.hpp [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+namespace mlpack {
+namespace rl {
+
+/**
+ * One step Sarsa worker.
+ *
+ * @tparam EnvironmentType The type of the reinforcement learning task.
+ * @tparam NetworkType The type of the network model.
+ * @tparam UpdaterType The type of the optimizer.
+ * @tparam PolicyType The type of the behavior policy. *
```
Comment (zoq) [inline src/mlpack/methods/reinforcement_learning/worker/one_step_sarsa_worker.hpp]: Can you remove the ``*`` at the end?

File: src/mlpack/methods/reinforcement_learning/worker/one_step_sarsa_worker.hpp [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+class OneStepSarsaWorker
+{
+ public:
+  using StateType = typename EnvironmentType::State;
+  using ActionType = typename EnvironmentType::Action;
+  using TransitionType = std::tuple<StateType, ActionType, double, StateType,
+      ActionType>;
+
+  /**
+   * @param updater The optimizer.
```
Comment (zoq) [inline src/mlpack/methods/reinforcement_learning/worker/one_step_sarsa_worker.hpp]: Can you add a method description here?

File: src/mlpack/methods/reinforcement_learning/worker/one_step_sarsa_worker.hpp [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+    #pragma omp atomic
+    totalSteps++;
+
+    pending[pendingIndex++] =
+        std::make_tuple(state, action, reward, nextState, nextAction);
+
+    if (terminal || pendingIndex >= config.UpdateInterval())
+    {
+      // Initialize the gradient storage.
+      arma::mat totalGradients(learningNetwork.Parameters().n_rows,
```
Comment (zoq) [inline src/mlpack/methods/reinforcement_learning/worker/one_step_sarsa_worker.hpp]: We should initialize ``totalGradients`` with zeros here.

File: src/mlpack/methods/reinforcement_learning/worker/one_step_sarsa_worker.hpp [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+        arma::colvec actionValue;
+        #pragma omp critical
+        {
+          targetNetwork.Predict(
+              std::get<3>(transition).Encode(), actionValue);
+        };
+        double targetActionValue = actionValue[std::get<4>(transition)];
+        if (terminal && i == pending.size() - 1)
+          targetActionValue = 0;
+        targetActionValue = std::get<2>(transition) +
```
Comment (zoq) [inline src/mlpack/methods/reinforcement_learning/worker/one_step_sarsa_worker.hpp]: Should we use an else case here? That way we can slightly simplify the expression if ``(terminal && i == pending.size() - 1)`` is true.

File: src/mlpack/methods/reinforcement_learning/worker/n_step_q_learning_worker.hpp [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+      const UpdaterType& updater,
+      const EnvironmentType& environment,
+      const TrainingConfig& config,
+      bool deterministic):
+      updater(updater),
+      environment(environment),
+      config(config),
+      deterministic(deterministic),
+      pending(config.UpdateInterval())
+  { reset(); }
```
Comment (zoq) [inline src/mlpack/methods/reinforcement_learning/worker/n_step_q_learning_worker.hpp]: Can you use Upper camel casing for all method names?

**Comment by ShangtongZhang** [comment]
Thanks for your feedback. Hope it's ready to merge now.

**Pull request merged**